component/prometheus/operator/podmonitors/crdmanager.go

```diff
@@ -0,0 +1,280 @@
+	DebugInfo        map[string]*DiscoveredPodMonitor
+	DiscoveryManager *discovery.Manager
+	ScrapeManager    *scrape.Manager
+
+	Opts      component.Options
+	Logger    log.Logger
+	Args      *Arguments
+	ConfigGen config_gen.ConfigGenerator
+}
+
+func NewCRDManager(opts component.Options, logger log.Logger, args *Arguments) *CRDManager {
+	return &CRDManager{
+		Opts:             opts,
+		Logger:           logger,
+		Args:             args,
+		DiscoveryConfigs: map[string]discovery.Configs{},
+		ScrapeConfigs:    map[string]*config.ScrapeConfig{},
+		DebugInfo:        map[string]*DiscoveredPodMonitor{},
+	}
+}
+
+func (c *CRDManager) Run(ctx context.Context) error {
+	c.ConfigGen = config_gen.ConfigGenerator{
+		Client: &c.Args.Client,
+	}
+
+	// Start prometheus service discovery manager
+	c.DiscoveryManager = discovery.NewManager(ctx, c.Logger, discovery.Name(c.Opts.ID))
+	go func() {
+		err := c.DiscoveryManager.Run()
+		if err != nil {
+			level.Error(c.Logger).Log("msg", "discovery manager stopped", "err", err)
+		}
+	}()
+
+	if err := c.runInformers(ctx); err != nil {
+		return err
+	}
+	level.Info(c.Logger).Log("msg", "informers  started")
+
+	// Start prometheus scrape manager.
+	flowAppendable := prometheus.NewFanout(c.Args.ForwardTo, c.Opts.ID, c.Opts.Registerer)
+	opts := &scrape.Options{}
+	c.ScrapeManager = scrape.NewManager(opts, c.Logger, flowAppendable)
+	defer c.ScrapeManager.Stop()
+	targetSetsChan := make(chan map[string][]*targetgroup.Group)
+	go func() {
+		err := c.ScrapeManager.Run(targetSetsChan)
+		level.Info(c.Logger).Log("msg", "scrape manager stopped")
+		if err != nil {
+			level.Error(c.Logger).Log("msg", "scrape manager failed", "err", err)
+		}
+	}()
+
+	// Start the target discovery loop to update the scrape manager with new targets.
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case m := <-c.DiscoveryManager.SyncCh():
+			targetSetsChan <- m
+		}
+	}
+}
+
+// runInformers starts all the informers that are required to discover PodMonitors.
+func (c *CRDManager) runInformers(ctx context.Context) error {
+	config, err := c.Args.Client.BuildRESTConfig(c.Logger)
+	if err != nil {
+		return fmt.Errorf("creating rest config: %w", err)
+	}
+
+	scheme := runtime.NewScheme()
+	for _, add := range []func(*runtime.Scheme) error{
+		v1.AddToScheme,
+	} {
+		if err := add(scheme); err != nil {
+			return fmt.Errorf("unable to register scheme: %w", err)
+		}
+	}
+
+	ls, err := c.Args.LabelSelector.BuildSelector()
+	if err != nil {
+		return fmt.Errorf("building label selector: %w", err)
+	}
+	for _, ns := range c.Args.Namespaces {
+		opts := cache.Options{
+			Scheme:    scheme,
+			Namespace: ns,
+		}
+
+		if ls != labels.Nothing() {
+			opts.DefaultSelector.Label = ls
+		}
+		cache, err := cache.New(config, opts)
+		if err != nil {
+			return err
+		}
+
+		informers := cache
+
+		go func() {
+			err := informers.Start(ctx)
+			if err != nil && ctx.Err() != nil {
```

Comment:
Is this correct? Did you mean `err != && ctx.Err() == nil`?

Reply:
not sure about this one, this is also used in other components like pod logs and k8s events similarly. It does make sense to me, could you please elaborate?

Reply:
looking at the [implementation](https://github.com/kubernetes-sigs/controller-runtime/blob/7982358244ab6ce4e3a919c1999dd79a5a4476c4/pkg/cache/internal/informers.go#L195) (I believe this is the one), seems like `Start` blocks until the the context is cancelled. Once the context is cancelled, `ctx.Err()` will always return an error (related to the cancellation).

Although this condition will never be triggered given this implementation, we want to log an error when `Start` returns an error and the error is not related to the cancellation, so your reasoning makes sense to me.

I will change this now (only in this component for now)

Reply:
I think that code is carried over from PodLogs. Your interpretation is as good as mine.